### DISCUSSION TITLE
How to capture network traffic from OpenThread network simulator?

### DISCUSSION THREAD

**Discussion by MartinsMihaeljans** [openthread/openthread#9194] [Q&A] [answered]
Hello, I followed this codelab - https://openthread.io/codelabs/openthread-network-simulator#0
What I would like to do is to capture simulated network traffic in Wireshark, is it possible? Or is there a built in capture that I can enable?
I already looked at instructions for packet sniffing with Pyspinel with extcap  - https://openthread.io/guides/pyspinel/install-pyspinel ,
but Wireshark does not recognize OpenThread sniffer if I have no hardware attached, is that correct?

**Comment by Irving-cl** [comment]
Hi @MartinsMihaeljans 

Each simulation node uses a udp socket to communicate. It encapsulates Thread frame into a udp frame on localhost. They will use ports `9001`, `9002`, etc. 
![Screenshot 2023-06-20 at 19 06 46](https://github.com/openthread/openthread/assets/12945188/d7753627-1a53-4cd3-980e-5304b9985747)

So you can capture all the nodes by capturing the loopback interface:
![Screenshot 2023-06-20 at 19 08 40](https://github.com/openthread/openthread/assets/12945188/f42003ee-ec59-42c9-a675-61259222a3e1)

And you can filter the Thread traffic by adding filter in wireshark like:
```
udp.port == 9001 || udp.port == 9002
```

**Reply by MartinsMihaeljans** [reply]
Thanks, I'll try and get back as soon as possible.

**Comment by MartinsMihaeljans** [comment]
@Irving-cl I succeeded with capturing the traffic, but now I have an additional question - can I recover the original headers and information, for example, IPv6 addressing? I understand that it might be contained in payload of the UDP packets, but is there a way to extract the information?
Here is an example of my capture:
<img width="578" alt="image" src="https://github.com/openthread/openthread/assets/128288466/b8bcacdb-9101-43d1-910f-c31619f303d4">

**Comment by MartinsMihaeljans** [comment]
@Irving-cl In addition to my previous question, I found that in this OTNS code there is a python script presumably for pcap creation. Could that be used for the capture of OpenThread packets?

Script is located here:
/home/administrator/src/openthread/tools/harness-sniffer/OT_Sniffer.py

Git been copied as in the codelab:
git clone https://github.com/openthread/openthread ~/src/openthread

**Comment by Irving-cl** [comment, marked as answer]
@MartinsMihaeljans 

Sorry I made a mistake. We cannot capture the packet in real time when `OPENTHREAD_SIMULATION_VIRTUAL_TIME` is used. And this option is required for OTNS. So we cannot capture the packet in real time with OTNS.

![Screenshot 2023-06-21 at 09 25 53](https://github.com/openthread/openthread/assets/12945188/d630a6c8-cc07-4364-a3a1-1150ce26e829)

While we cannot capture it in real time, there will be a pcap file generated. After running `OTNS`, you can find `current.pcap` under the same directory as `ot-cli-xxx`:
```
(base) irvingcl@irvingcl1:~/Documents/git/openthread/build/simulation/examples/apps/cli$ ls
CMakeFiles  cmake_install.cmake  CTestTestfile.cmake  current.pcap  ot-cli-ftd  ot-cli-mtd  ot-cli-radio  otns_0.replay  tmp
```
You can directly open it with wireshark:
```
wireshark current.pcap
```

You don't have to open it after shutting down OTNS. But it cannot update in real time as I tried.

---

> /home/administrator/src/openthread/tools/harness-sniffer/OT_Sniffer.py

This is a sniffer to capture real Thread packets. It's not related with simulation.

**Reply by MartinsMihaeljans** [reply]
Super, thank you @Irving-cl this works for me as well!
